### PR TITLE
Enhanced Permissions

### PR DESCRIPTION
**Pull request by steve-keep**
Enhance the current permission system with the following:

- [x] Implement implicit deny/allow
- [x] Search filters for read access
- [x] Support DeleteMany
- [x] Search Inline fragments for access
- [x] Search Fragment spread for access
- [x] Search Nested filters for access
- [x] Allow ACL function to return a boolean
- [x] Apply GraphQL Armor - Set max-depth to 6

https://app.clickup.com/t/2704425/ID-6138

### CODE REVIEW

File: src/packages/end-to-end/src/__tests__/api/auth/acl/nested-filters.test.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+const graphweaver = new Graphweaver({
+	resolvers: [AuthResolver, TrackResolver, AlbumResolver, ArtistResolver],
+	apolloServerOptions: {
+		plugins: [authApolloPlugin(async () => user)],
+	},
+});
+
+let token: string | undefined;
+
+describe('ACL - Fragments', () => {
```
Comment (thekevinbrown) [inline src/packages/end-to-end/src/__tests__/api/auth/acl/nested-filters.test.ts:140]: Are these `fragments` we're testing here?

File: src/packages/server/package.json [hunk 2]
```diff
@@ -22,6 +22,8 @@
 	"dependencies": {
 		"@apollo/server": "4.9.3",
 		"@as-integrations/aws-lambda": "2.0.1",
+		"@escape.tech/graphql-armor": "2.4.0",
+		"@escape.tech/graphql-armor-types": "0.5.0",
```
Comment (thekevinbrown) [inline src/packages/server/package.json:26]: Suspect this is a dev dependency.

**Review by thekevinbrown — APPROVED** [review]
Looking great! A few small things that I think could be better, but I'll leave it to you to take or leave as you see fit.

File: src/packages/auth/src/auth-utils.ts [hunk 3]
```diff
@@ -92,32 +93,44 @@ const permissionsErrorHandler = (error: any) => {
...
+		if (!authContext) {
+			throw new Error('Authorisation context provider not initialised');
+		}
+
+		// Let's resolve the filter functions and check if any of them return false or undefined
+		const filterFunctions = acv.map(async (value) => value(authContext as TContext));
+		const resolvedFilterFunctions = await Promise.all(filterFunctions);
+
+		for (const filterValue of resolvedFilterFunctions) {
+			if (filterValue === false || filterValue === undefined) {
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/packages/auth/src/auth-utils.ts]: On this one, I think if the function returns `{ cow: "moo" }` then we need to allow it through. 

It means that we have an input filter that needs to be checked and sent to the data provider.

The user is allowed access but only with that filter. Would you agree?

[the remark this answers is not in the document]
Comment (thekevinbrown) [reply src/packages/auth/src/auth-utils.ts]: Ah, yup, forgot about that use case, I do agree. What if they return `null` or `"false"`?

- Would the first imply no filter?
- And the second seems like a whoopsie to me.

[the remark this answers is not in the document]
Comment (steve-keep) [reply src/packages/auth/src/auth-utils.ts]: Yeah, you are right, I now have this check:

https://github.com/exogee-technology/graphweaver/pull/309/commits/d8acf7f2affe01c76a1c5ecdc01e9a9dafbe4de3#diff-18f8a9cbeb1172bc9921cd23fd047e60ba670b06d682317b258ea605ab3c67d9R135

**Pull request merged**